modules/SofaBoundaryCondition/FixedPlaneConstraint.h

```diff
@@ -63,120 +70,88 @@ class FixedPlaneConstraint : public core::behavior::ProjectiveConstraintSet<Data
+    void addConstraint(int index);
+    void removeConstraint(int index);
 
+protected:
+    FixedPlaneConstraint();
+    ~FixedPlaneConstraint();
 
+    FixedPlaneConstraintInternalData<DataTypes> data;
+    friend class FixedPlaneConstraintInternalData<DataTypes>;
 
-        void applyDestroyFunction(unsigned int /*index*/, value_type& /*T*/);
+    /// Forward class declaration, definition is in the .inl
+    class FCPointHandler;
 
+    /// Handler for subset Data
+    FCPointHandler* m_pointHandler {nullptr};
 
-        bool applyTestCreateFunction(unsigned int /*index*/,
-                const sofa::helper::vector< unsigned int > & /*ancestors*/,
-                const sofa::helper::vector< double > & /*coefs*/);
-    protected:
-        FixedPlaneConstraint<DataTypes> *fc;
-    };
+    /// whether vertices should be selected from 2 parallel planes
+    bool m_selectVerticesFromPlanes {false};
 
-protected:
     /// Pointer to the current topology
-    sofa::core::topology::BaseMeshTopology* topology;
-
-    /// Handler for subset Data
-    FCPointHandler* pointHandler;
-
-    bool isPointInPlane(Coord p)
-    {
-        Real d=dot(p,direction.getValue());
-        if ((d>dmin.getValue())&& (d<dmax.getValue()))
-            return true;
-        else
-            return false;
-    }
+    BaseMeshTopology* m_topology {nullptr};
+
+    ////////////////////////// Inherited attributes ////////////////////////////
+    /// https://gcc.gnu.org/onlinedocs/gcc/Name-lookup.html
+    /// Bring inherited attributes and function in the current lookup context.
+    /// otherwise any access to the base::attribute would require
+    /// the "this->" approach.
+    using ProjectiveConstraintSet<DataTypes>::mstate;
+    using ProjectiveConstraintSet<DataTypes>::getContext;
+
+    /// These two are implemented depending on the templates
+    bool isPointInPlane(Coord p) const ;
+
+    /// These two are implemented depending on the templates
+    template<class T>
+    void projectResponseImpl(const MechanicalParams* mparams, T& dx) const ;
 };
 
-#ifndef SOFA_FLOAT
-template <> template <class DataDeriv>
-void FixedPlaneConstraint<defaulttype::Rigid3dTypes>::projectResponseT(const core::MechanicalParams* mparams, DataDeriv& /*res*/);
-
-template <>
-bool FixedPlaneConstraint<defaulttype::Rigid3dTypes>::isPointInPlane(Coord /*p*/);
-#endif
-
-#ifndef SOFA_DOUBLE
-template <> template <class DataDeriv>
-void FixedPlaneConstraint<defaulttype::Rigid3fTypes>::projectResponseT(const core::MechanicalParams* mparams, DataDeriv& /*res*/);
-
-template <>
-bool FixedPlaneConstraint<defaulttype::Rigid3fTypes>::isPointInPlane(Coord /*p*/);
-#endif
-
-#if defined(SOFA_EXTERN_TEMPLATE) && !defined(SOFA_COMPONENT_PROJECTIVECONSTRAINTSET_FIXEDPLANECONSTRAINT_CPP)
-#ifndef SOFA_FLOAT
+#if !defined(SOFA_COMPONENT_PROJECTIVECONSTRAINTSET_FIXEDPLANECONSTRAINT_CPP)
```

Comment:
If I understand correctly, this #ifdef is there to make sure we do not include the extern templates in the _FixedPlaneConstraint.cpp_ file, since we need to instantiate these at least once (here it would be in _FixedPlaneConstraint.o_).

Correct me if I'm wrong, but I don't think this #ifdef is needed.

`extern template class ...` basically says to the compiler: do not instantiate these templates since they will be defined somewhere else (here it would be in _FixedPlaneConstraint.o_). This will save some compilation time as the other .o files will be able to simply link to _FixedPlaneConstraint.o_ to get these template instances.

In _FixedPlaneConstraint.cpp_, you would have
`template class ...` that basically says to the compiler, no mater what I said before ( ex. the `extern template class...` in the header), force the instantiation of these templates in this object file (_FixedPlaneConstraint.o_)

This means that in other object files (.cpp) that include the header with the `extern template class ...`, they will not implement the template, but in the _FixedPlaneConstraint.cpp_, the instantiation will be forced (we need it to be somewhere)

Hence, no need for the #ifdef

Reply:
Thanks for that very clear explanation!
@epernod Could you remember the case where you had a problem removing such #ifdef ?

Reply:
I was thinking the same as @jnbrunet and this is what the standard says:
https://en.cppreference.com/w/cpp/language/class_template

I have a vague remembering compilation problem that I think were cause by the presence of explicit specialization and got the same gcc error as in this thread: https://stackoverflow.com/questions/21112148/specialization-of-member-function-template-after-instantiation-error-and-order/21112421